### ISSUE TITLE
Offer general classes like `BayesianQuadrature` that users can subclass to implement their own scheme.

### ISSUE THREAD

**Issue by JonathanWenger** [probabilistic-numerics/probnum#25]
Make general classes implementing PN methods public so the user can use the general interface (e.g. `bayesquad`) or specific method (e.g. `VanillaBayesianQuadrature`).

```python
class _BayesianQuadrature(abc.ABC):
    """
    An abstract base class for Bayesian Quadrature methods.
    """

    def __init__(self, func, func0):
        """
        Parameters
        ----------
        func : function
            Symbolic representation of the function to be integrated.
        func0 : RandomProcess
            Stochastic process modelling function to be integrated.
        """
        self.func = func
        self.func0 = func0

    def integrate(self, domain, **kwargs):
        """
        Integrate the function ``func``.

        Parameters
        ----------
        kwargs

        Returns
        -------

        """
        raise NotImplementedError
```

**Comment by pnkraemer** [comment]
I can help, too!

**Comment by pnkraemer** [comment]
I think it would be cool to eventually have e.g. a jupyter notebook that gives examples how to write extensions to some modules. There will surely be easy candidates (at least in the diffeq module I can think of a few already) and this might be useful for eventual contributions.

**Comment by JonathanWenger** [comment]
Great Idea. We could have an Advanced section in the examples in the documentation.

**Comment by pnkraemer** [comment]
Sounds great! I like the idea of having different levels of examples "getting started", "exploring more features" and "contributing to the code" and each section shows examples of different levels of difficulties

**Comment by JonathanWenger** [comment]
Implemented for the existing classes by a58e998.
